src/v/storage/backlog_controller.cc

```diff
@@ -0,0 +1,167 @@
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "storage/backlog_controller.h"
+
+#include "config/configuration.h"
+#include "prometheus/prometheus_sanitize.h"
+#include "ssx/sformat.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/metrics.hh>
+#include <seastar/core/reactor.hh>
+
+namespace storage {
+
+backlog_controller_config::backlog_controller_config(
+  double kp,
+  double ki,
+  double kd,
+  int64_t normalization_factor,
+  int64_t sp,
+  int initial_shares,
+  std::chrono::milliseconds interval,
+  ss::scheduling_group sg,
+  ss::io_priority_class iop,
+  int min,
+  int max)
+  : proportional_coeff(kp)
+  , integral_coeff(ki)
+  , derivative_coeff(kd)
+  , normalization_factor(normalization_factor)
+  , setpoint(sp)
+  , initial_shares(initial_shares)
+  , sampling_interval(interval)
+  , scheduling_group(sg)
+  , io_priority(iop)
+  , min_shares(min)
+  , max_shares(max) {}
+
+backlog_controller::backlog_controller(
+  std::unique_ptr<sampler> sampler,
+  ss::logger& logger,
+  backlog_controller_config cfg)
+  : _sampler(std::move(sampler))
+  , _log(logger)
+  , _kp(cfg.proportional_coeff)
+  , _ki(cfg.integral_coeff)
+  , _kd(cfg.derivative_coeff)
+  , _norm(cfg.normalization_factor)
+  , _sampling_interval(cfg.sampling_interval)
+  , _scheduling_group(cfg.scheduling_group)
+  , _io_priority(cfg.io_priority)
+  , _setpoint(cfg.setpoint / _norm)
+  , _current_shares(cfg.initial_shares)
+  , _min_shares(cfg.min_shares)
+  , _max_shares(cfg.max_shares) {}
+
+ss::future<> backlog_controller::start() {
+    _current_backlog = co_await _sampler->sample_backlog() / _norm;
+    _prev_error = _setpoint - _current_backlog;
+    _sampling_timer.set_callback([this] {
+        (void)ss::with_gate(_gate, [this] {
+            return update().then([this] {
+                if (!_gate.is_closed()) {
+                    _sampling_timer.arm(_sampling_interval);
+                }
+            });
+        });
+    });
+    _sampling_timer.arm(_sampling_interval);
+}
+
+void backlog_controller::update_setpoint(int64_t v) {
+    vlog(_log.info, "new setpoint value: {}, {}", v, _ki);
+    _setpoint = v;
+}
+
+ss::future<> backlog_controller::stop() {
+    if (_sampling_timer.armed()) {
+        _sampling_timer.cancel();
+    }
+    return _gate.close();
+}
+
+ss::future<> backlog_controller::update() {
+    _current_backlog = co_await _sampler->sample_backlog() / _norm;
+    auto current_err = _setpoint - _current_backlog;
+
+    accumulate_error(current_err);
+    auto update = static_cast<double>(current_err);
+    if (_ki != 0.00) {
+        update += _ki * static_cast<double>(_error_integral);
+    }
+    if (_kd != 0.00) {
+        update += _kd * static_cast<double>(current_err - _prev_error);
+    }
+    update *= _kp;
```

Comment:
Can you double check your algo here, I expected something more like:
```suggestion
    auto update = _kp * static_cast<double>(current_err);
    if (_ki != 0.00) {
        update += _ki * static_cast<double>(_error_integral);
    }
    if (_kd != 0.00) {
        update += _kd * static_cast<double>(current_err - _prev_error);
    }
```

Reply:
both versions are correct, in this case all the coefficients are 'scaled' with proportional terms. I can change it so all of the terms will be independent.

Reply:
when I think of it now, I think it is more convenient to scale integral and derivative terms with proportional  coefficient, this way we keep the terms in the same 'range of values' additionally we would have to remember when `I` & `D` terms should be negative.